stellargraph/layer/hinsage.py

```diff
@@ -310,13 +313,20 @@ def eval_neigh_tree_per_layer(input_tree):
             for layer, dim in enumerate([self.input_dims] + layer_sizes)
         ]
 
+        # Activation function for each layer
+        if activations is None:
+            activations = ["relu"] * (self.n_layers - 1) + ["linear"]
+        elif len(activations) != self.n_layers:
+            raise ValueError(
```

Comment:
I'd perhaps raise an IndexError here.
Also in GraphSAGE line 732.

Reply:
This one I'm not sure about - various code raises ValueError, TypeError, etc. Sometimes I can see the rationale, e.g. a str was passed where an int was required - maybe TypeError is sensible. Other times, it doesn't seem so clear-cut, so I have opted generically for ValueError, sort of as a proxy to Java's IllegalArgumentException. I'm open to discussion on this point; maybe we can have (or already have) some clear guidelines.

Reply:
Just to capture my thinking, suppose the input requirement was for a list of exactly N integers between 0 and M, but suppose the actual input given was a list comprised of N-3 integers between 0 and M, one integer > M, and one string. Then:
* if length is checked first, we might get an IndexError;
* if type is checked first, we might get a TypeError;
* if value is checked first, we might get a ValueError.
Hence my reasoning that it might be easier (including for any try/catch wrapper) to just have a single exception that indicates that not all conditions have been satisfied.

Reply:
Yes, it's rather ambiguous in Python what error to raise, especially when there are multiple things wrong in the input.  Nonetheless, regardless of which order things are checked as long as we state what the error is the user will get a more informative message.